src/emqx_protocol.erl

```diff
@@ -876,17 +878,18 @@ check_subscribe(TopicFilters, PState = #pstate{zone = Zone}) ->
             {error, TopicFilter1}
     end.
 
-check_sub_acl(TopicFilters, #pstate{credentials = #{is_superuser := IsSuper}, enable_acl = EnableAcl})
-        when IsSuper orelse (not EnableAcl) ->
+check_sub_acl(TopicFilters, #pstate{credentials = #{is_superuser := IsSuper}})
+        when IsSuper ->
     {ok, TopicFilters};
-check_sub_acl(TopicFilters, #pstate{credentials = Credentials}) ->
+check_sub_acl(TopicFilters, #pstate{zone = Zone, credentials = Credentials}) ->
+    EnableAcl = emqx_zone:get_env(Zone, enable_acl, false),
     lists:foldr(
-      fun({Topic, SubOpts}, {Ok, Acc}) ->
-              case emqx_access_control:check_acl(Credentials, subscribe, Topic) of
-                  allow -> {Ok, [{Topic, SubOpts}|Acc]};
-                  deny  ->
-                      {error, [{Topic, SubOpts#{rc := ?RC_NOT_AUTHORIZED}}|Acc]}
-              end
+      fun({Topic, SubOpts}, {Ok, Acc}) when EnableAcl ->
+              AllowTerm = {Ok, [{Topic, SubOpts}|Acc]},
+              DenyTerm = {error, [{Topic, SubOpts#{rc := ?RC_NOT_AUTHORIZED}}|Acc]},
+              do_acl_check(subscribe, Credentials, Topic, AllowTerm, DenyTerm);
+         (TopicFilter, Acc) ->
```

Comment:
This clause is never to be executed. It should be deleted.

Reply:
Nope